### PR TITLE
[sw/dif_keymgr] Key Manager DIF Library Header

### PR DESCRIPTION
**Pull request by alphan**
This change adds the header of the key manager DIF library that defines its public interface.

Relevant docs:
* [Key Manager HWIP Technical Specification](https://docs.opentitan.org/hw/ip/keymgr/doc)
* [Identities and Root Keys](https://docs.opentitan.org/doc/security/specs/identities_and_root_keys/)

Signed-off-by: Alphan Ulusoy <alphan@google.com>

### CODE REVIEW

File: sw/device/lib/dif/dif_keymgr.c
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.c:9]: I would drop the #error, since it makes it so this library can't be built. I think it's fine to just have an empty file so that the header is built and checked.

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:78]: Nit: OK -> Ok

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:141]: Pwrmgr? =P

I would drop this and add it only if you wind up needing it in the impl.

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:293]: Thoughts: Is "initialize" a good word here? It feels confusable with "init". I don't know what the best word is, but I think something like "prime" or "seed" might be more appropriate.

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:318]: Do you think this should use the `transaction_t` naming style?

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:29]: Optional: to avoid what are technically unnecessary macros, I think it would be ok to write something like
```c
enum {
  kDifKeymgrSwBindingWords = 4,
};
```

Me and @lenary have thought about doing this to avoid unnecessary macro names I don't think we should block on it.

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:357]: This should be called `dif_keymgr_lock()`, I think?

(With an `is_locked()` companion.)

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:368]: Note; the parens are unnecessary.

I would consider making these just regular bit indices, so that they can be used with `bitfield_bit32_*()`. `otp_ctrl` has this kind of API (which makes the indices usable as array indices, as well). You don't *need* it for this API but it would be nice if all the DIFs were consistent on this front.

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:415]: Do these *need* to be queried together?

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:308]: Is there a particular reason this isn't a byte array? I assume the alignment isn't tremendously important.

File: sw/device/lib/dif/meson.build
Comment (mcy) [inline sw/device/lib/dif/meson.build:190]: Can you uncomment this so the header gets compiled as part of CI?

File: sw/device/lib/dif/dif_keymgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_keymgr.h:376]: Nit: I think we're pretty consistent in not all-caps-ing acronyms in constants (so KMAC -> Kmac). I the style guide actually mandates this spelling.

**Review by mcy — COMMENTED** [review]
Hi Alphan, I've done a first-pass review on the code so far. LMK what you think. =)

**Comment by alphan** [comment]
Thanks for the quick review @mcy! It looks like I accidentally created this as a ready-for-review PR from github cli, apologies for the noise :) I'll ping you once I hear from HW folks and update the PR.

**Comment by mcy** [comment]
Haha, sounds good. =)

**Comment by lenary** [comment]
Garret and I have scheduled a Design Review of this for Monday, 23rd November 2020.

File: sw/device/lib/dif/dif_keymgr.c
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.c:9]: This is based on @lenary's comment [here](https://github.com/lowRISC/opentitan/pull/3142#discussion_r493444840). I don't have a strong preference either way.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:294]: I had the same concern but thought prime/seed might be slightly inaccurate because this value is not used for computing the secret value of the following states.

This function can also be dropped but I am inclined to keep it until the HW design is finalized. Leaving this open for others to share their thoughts.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h]: This is not a parameter of advance_state anymore.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:392]: I considered this option too, but chose this name to mostly align with the tech spec. Also, since we already have a getter for the status, I thought is_locked would be redundant.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:403]: I definitely agree with the consistency argument. I don't have a strong preference for either but IMHO defining these as masks make it a bit more obvious how these are supposed to be used but OTOH eliminates the possibility of using `bitfield_bit32_*()` by callers, which may hurt readability.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:450]: Not really, but currently HW handles initialization in a different way for the sake of efficiency (not as an operation but as a separate command and reports its status in WORKING_STATE.STATE instead of OP_STATUS.STATUS). I tried to abstract this away (HW may change to do the same). Getting the status would require reading both registers anyway so I thought it would be convenient to keep everything related in a single struct.

File: sw/device/lib/dif/meson.build
Comment (alphan) [reply sw/device/lib/dif/meson.build:190]: Related to [this](https://github.com/lowRISC/opentitan/pull/4256#discussion_r526987407).

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [inline sw/device/lib/dif/dif_keymgr.h:358]: I initially had this as a part of `dif_keymgr_advance_state()` but based on our discussions w/ @tjaychen and @moidx we will probably need a way to set and lock state inputs before advancing state due to lack of entropy. `dif_keymgr_state_params_t` does not hold the parameters of every state because every stage in the secure boot chain will set the parameters of the next state.

Regarding locking: There is no need to postpone locking but we may want to handle it separately by defining `*_lock` and `*_is_locked` depending on how we want to handle hardening.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [inline sw/device/lib/dif/dif_keymgr.h:105]: I'm aware that this is currently missing a `Busy` constant. I need to confirm the behavior w/ the HW team.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [inline sw/device/lib/dif/dif_keymgr.h:309]: `RomExtSecurityDescriptor` is currently a 128b value in the tech spec but based on our discussions it is expected to be 256b. We may not need these constants if the software binding value ends up using the same size. Using a union in `dif_keymgr_state_params_t ` instead of this is also another option.

**Comment by gkelly** [comment]
Just a gentle reminder, reviews before header reviews should probably be avoided in order to reduce the toil for the header author.

File: sw/device/lib/dif/dif_keymgr.md
Comment (lenary) [inline sw/device/lib/dif/dif_keymgr.md:24]: Nit: Can you add spaces in the `smoke` lines? I see you did a search/replace for sanity, sadly they have different lengths.

File: sw/device/lib/dif/dif_keymgr.h
Comment (gkelly) [inline sw/device/lib/dif/dif_keymgr.h:204]: @tjaychen For the `Init_wip` state: why does it exist, and exposed to software? Is there some special treatment of the device we need to do while it's active? The other state transitions are asynchronous and aren't visible to software, so I'm curious about the difference.

**Review by gkelly — COMMENTED** [review]
Thanks for writing this up, @alphan! @lenary and I did a DIF review of this change. We noticed that there's some strangeness around the hardware interface, an we've tagged the HWEs on those questions. We'd like to see some separation in the status/state accessors so that a user can query one at a time. We'd like to hold off on the implementation until we've seen answers to the hardware concerns we raised.

Thanks!

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:309]: `RomExtSecurityDescriptor`is now implemented through software binding (#4486).

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:134]: Error IRQ is removed in #4513.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:449]: Yes, good catch! But we don't have this struct anymore (after defining individual getters for the state and status codes).

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:204]: `Init_wip` was removed in #4467.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:372]: IIUC, `OP_STATUS` (rw1c) will be `WIP` during an operation even if the software clears it. Once the operation ends, the software will read `DONE_SUCCESS` or `DONE_ERROR` and clearing those will reset `OP_STATUS` to `Idle`. @tjaychen, is this correct?

**Comment by tjaychen** [comment]
yes that's right.  At least that's how it was intended, can't say there
aren't bugs of course :)

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:358]: I decided to remove this function because there is a single register for the binding value now.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:105]: I don't think we need a separate enum for busy.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:399]: Yes, you're right. It looks like I named the error values looking at the description in the spec instead of the register table. `kDifKeymgrStatusInvalidInput` should have been `kDifKeymgrStatusInvalidKmacCommand`.

Regarding translation: Yes, there is a one-to-one mapping between the bits of the `ERROR_CODE` register and the first four bits of `dif_keymgr_status_codes_t`. The fifth bit is defined as `OP_STATUS != WIP`. Other values of `OP_STATUS` is used to determine if we need to check the `ERROR_CODE` register or not.

**Comment by alphan** [comment]
Thanks for reviewing this @gkelly, @lenary, and @mcy. @tjaychen, thanks for your time and the improvements in the software interface! Rebased and pushed a new commit (just to highlight the differences, will be squashed before merging) with the following changes (mostly in response to the changes in HW):
- Addressed reviewers' comments.
- Added `dif_keymgr_clear_sideload_keys()`.
- Added `dif_keymgr_configure()` and `dif_keymgr_config_t` to support reseed interval.
- Removed error interrupt.
- Removed `dif_keymgr_initialize_state()`
- Removed `RomExtSecurityDesc` related definitions, since binding value is used for both now.
- Removed `dif_keymgr_set_state_params()` and moved next state parameters to `dif_keymgr_advance_state()`.
- Updated comments as needed.

File: sw/device/lib/dif/dif_keymgr.h
Comment (moidx) [inline sw/device/lib/dif/dif_keymgr.h:303]: Sorry, late to the review party, would it be ok to list this as `uint32_t binding_value[4]`?

Comment (moidx) [inline sw/device/lib/dif/dif_keymgr.h:303]: @alphan, @tjaychen 

If we are using `binding_value` as the `ROMExtDescriptor`,  would it be possible to make the binding value `256B` so that we can use hash values if needed?

File: sw/device/lib/dif/dif_keymgr.h
Comment (moidx) [inline sw/device/lib/dif/dif_keymgr.h:485]: Is it possible to make this `uint32_t salt[4]`?

File: sw/device/lib/dif/dif_keymgr.h
Comment (moidx) [inline sw/device/lib/dif/dif_keymgr.h:529]: `uint32_t value[2][8]`

File: sw/device/lib/dif/dif_keymgr.h
Comment (moidx) [inline sw/device/lib/dif/dif_keymgr.h:485]: @tjaychen, @alphan 

Since we are combining ID and salt (I assume), is it possible to make this 256B?

**Review by moidx — COMMENTED** [review]
Added a few comments.

**Comment by tjaychen** [comment]
hey all, regarding increasing the width of sw_binding and salt to `256b`, that is fine with me.
But can we clear it with @moidx to make sure everything is inline?

Then we should first push a separate PR to update the identities security doc to make sure there is no confusion later.

**Comment by mateenfaisal** [comment]
What is blocking this PR? pinging reviewers @silvestrst, @mcy....

**Comment by silvestrst** [comment]
> What is blocking this PR? pinging reviewers @silvestrst, @mcy, @moidx, @gkelly

I haven't reviewed this change, can do next week, otherwise if others think it is ready, they should approve and merge it.

**Review by gkelly — APPROVED** [review]
Please follow update to @moidx's new types before submitting. Thanks @alphan!

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:303]: Done.

Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:303]: Done.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:485]: Done.

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:529]: Done

File: sw/device/lib/dif/dif_keymgr.h
Comment (alphan) [reply sw/device/lib/dif/dif_keymgr.h:485]: Done.

**Comment by alphan** [comment]
> hey all, regarding increasing the width of sw_binding and salt to `256b`, that is fine with me.
> But can we clear it with @moidx to make sure everything is inline?
> 
> Then we should first push a separate PR to update the identities security doc to make sure there is no confusion later.

I agree, created #5193 and #5194 to track doc and rtl changes.

**Pull request merged**
